crates/driver/src/domain/competition/solution/settlement.rs

```diff
@@ -77,3 +80,51 @@ impl Verified {
         self.inner.boundary
     }
 }
+
+/// Gas parameters associated with a settlement.
+#[derive(Clone, Copy, Debug)]
+pub struct Gas {
+    /// The gas estimate, in gas units, for executing a settlement transaction.
+    pub estimate: eth::Gas,
+    /// The gas limit, in gas units, for a settlement transaction. This is
+    /// computed by adding a buffer to the gas estimate to allow for small
+    /// variations in the actual gas that gets used.
+    pub limit: eth::Gas,
+    /// The maximum fee per unit of gas for a given settlement.
+    pub price: eth::MaxFeePerGas,
+}
+
+impl Gas {
+    /// Computes settlement gas parameters given estimates for gas and gas
+    /// price.
+    pub fn new(estimate: eth::Gas, price: eth::GasPrice) -> Self {
+        // Compute an upper bound for `max_fee_per_gas` for the given
+        // settlement. We multiply a fixed factor of the current base fee per
+        // gas, which is chosen to be the maximum possible increase to the base
+        // fee per gas over 10 blocks.
+        const MAX_FEE_FACTOR: f64 = 3.25;
+        let price =
+            eth::U256::from_f64_lossy(eth::U256::to_f64_lossy(price.base.into()) * MAX_FEE_FACTOR)
+                .into();
```

Comment:
This is for an upper-bound, I don't think loss-less behavior is important here (as we will pay the `base_fee_per_gas` regardless of this value).

Reply:
When we execute a settlement, we specify that the `max_fee_per_gas` cannot exceed `settlement.gas.price`:

https://github.com/cowprotocol/services/blob/ef9c8d278dd54f1f96d42b4c1b64634c67fd7b9e/crates/driver/src/boundary/mempool.rs#L119-L130

And in the submission logic:

https://github.com/cowprotocol/services/blob/ef9c8d278dd54f1f96d42b4c1b64634c67fd7b9e/crates/solver/src/settlement_submission/submitter.rs#L164-L171

This means that, even if we run into a situation where:
1. We get an initial gas price estimate with `base: 10, max: 20`
2. We compute the maximum we want to pay as `base * factor ~= 33`
3. Start submission loop capping gas price to `33`
    - If the estimator gives us `base: 20, max: 40`, we will use `max: 33` (and be guaranteed to have enough balance for it).
    - If the estimator gives us `base: 100, max: 200`, we will fail the submission as the base fee surpassed our cap (this can happen, for example, on Gnosis Chain where the block time is shorter, because of exponents).
    
Basically this change makes us no longer rely on the estimator's `max_fee_per_gas` but to compute our own based on the current base fee and how much we reasonably want to over-pay what it currently is.

Reply:
I see. Thanks for the explanation